### PR TITLE
Add reload_after_ready flag to Reload (Draft)

### PR DESCRIPTION
**Pull request by rokoroku**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | ~yes~ no
| Fixed tickets | #3429 
| License       | MIT
| Doc PR        | https://github.com/pm2-hive/pm2-hive.github.io/pulls

This PR is a draft to achieve a feature I suggested in #3429 

I'm not sure how the flag should be like.. there are many options
1. `preserveOldWhenFail` - only resotre old process when new process crashes
2. `preserveOldWhenNoReady` - only resotre old process when new process does not send 'ready'
3. `stopWhenFail` - only stop the process when new process crashes: this is a BREAKING CHANGE that affects default behavior
... and more?

It's not a completed PR. I just want to ask the way from here.

### CODE REVIEW

File: lib/God/Reload.js [hunk 1]
```diff
@@ -157,7 +157,22 @@ function hardReload(God, id, wait_msg, cb) {
...
+          if (err) return cb(err);
+
+          // Restore old worker
+          God.clusters_db[id] = old_worker;
+          delete God.clusters_db[t_key];
+
+          old_worker.pm2_env.pm_id = id;
+          old_worker.pm_id = id;
+
+          cb(null, proc);
```
Comment (rokoroku) [inline lib/God/Reload.js]: Maybe I can throw an error here to tell reload process has been failed.

**Comment by wallet77** [comment]
I think the best is between option 1 & 2.

Ideally we should find the best behavior to ensure 0 downtime.
Something like : 
1 - user try to reload or restart
2 - create a new worker (for each existing worker)
3 - if new worker fail or is not ready after a certain delay -> destroy it and keep old one
   - if new worker is ready, disable old one and redirect load to the new one + wait old worker to process old connections (see https://github.com/Unitech/pm2/issues/3407)

What do you think about that ?

**Comment by rokoroku** [comment]
This PR is exactly for what you said.

However, I don't think it's required to copy previous files to another folder since **build process** is not the main feature handled by PM2. 

I want to know which flags should I add to this PR, and which additional behavior should added to this PR .

**Comment by pinkasey** [comment]
@Unitech 
Why was this PR closed?
This feature is VERY MUCH IN NEED.

The behavior of "don't reload process X before process X-1 is done reloading _**SUCCESSFULLY**_" - is what we all want.
anyone who uses `pm2 reload` wants that.
No one wants their server to go completely down, during a bad deployment (deployment with a bug that prevents server from booting).

Won't anyone pick the glove from here?

**Pull request closed without merging**